holoscribe/src/cli.rs

```diff
@@ -1 +1,142 @@
+use std::num::ParseIntError;
 
+use clap::Parser;
+use regex::Regex;
+use thiserror::Error;
+
+#[derive(Error, Debug, PartialEq)]
+pub enum CliError {
+    #[error("Invalid int specification for size")]
+    InvalidSizeInt(ParseIntError),
+    #[error("Invalid size specification, please use the format: width[xheight](mm|cm|m)")]
+    InvalidSize,
+}
+
+#[derive(Parser)]
+#[command(author, version, about, long_about = None)]
+#[command(propagate_version = true)]
+/// A command line utility that take an .obj model and produce a .svg representations of lines to etch
+/// to create a holographic image of that model
+pub struct Args {
+    /// Input file. Expects a .obj file
+    #[arg(short, long)]
+    pub input: String,
+
+    /// Output file. Expects a .svg file
+    #[arg(short, long)]
+    pub output: String,
+
+    /// The size of the image you which to create. Requires the format: width[xheight](mm|cm|m)
+    #[arg(short, long, value_parser=parse_size)]
+    pub size: Size,
+
+    /// The density of lines etched per mm. Defaults to 1.
+    #[arg(long, default_value_t = 1)]
+    pub stroke_density: usize,
+}
+
+/// Represents a size in millimeters
+#[derive(Debug, Clone, PartialEq)]
+pub struct Size {
+    pub width: usize,
+    pub height: usize,
+}
+
+fn parse_size(arg: &str) -> Result<Size, CliError> {
+    let re = Regex::new(r"^(\d+)x?(\d*)(mm|cm|m)$").unwrap();
+    if let Some(cap) = re.captures(arg) {
+        // We always expect to match the first group (width)
+        let width_str = cap.get(1).unwrap().as_str();
+        let width = width_str
+            .parse::<usize>()
+            .map_err(|e| CliError::InvalidSizeInt(e))?;
+
+        // The second group (height) is optional, without it just use width
+        let height = if let Some(height) = cap.get(2) {
+            let height_str = height.as_str();
+            if height_str.len() > 0 {
+                height
+                    .as_str()
+                    .parse::<usize>()
+                    .map_err(|e| CliError::InvalidSizeInt(e))?
+            } else {
+                width
+            }
+        } else {
+            width
+        };
+
+        // We always expect to have a unit
+        let unit = cap.get(3).unwrap().as_str();
```

Comment:
would be nice with a error here specifying what went wrong :)

Reply:
Yeah, I'll write a panic message with more detail - but I don't think I want to return an error to the user since this would indicate a bug in the regex library

Reply:
I would be of a mind to just accept a default unit of mm if the user doesn't specify one

Reply:
I can see an argument for both but I'd learn towards requiring units. I like that it forces the user to think about the fact that the thing they are creating will have physical dimensions - especially because the output is a SVG, which doesn't normally have any.

Reply:
Or maybe I've missunderstood how the regex matching works!

Reply:
I probably have

Reply:
Yeah - if the user input doesn't have one of the 3 unit options, the regex should not match at all and it will return the error in the else statement.

Reply:
Aaaaaaaa